content/blog/2020-09-17-delivering-learnr-tutorials-in-a-package/index.Rmarkdown

```diff
@@ -0,0 +1,322 @@
+tags:
+  - rmarkdown
+  - teach
+  - learnr
+description: |
+  A step by step guide to create a package for a learnr tutorial.
+photo:
+  url: https://unsplash.com/photos/sxNt9g77PE0
+  author: Erda Estremera
+---
+
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(fig.align = "center", out.width = "50%")
+```
+
+## A package for the learnr occasion
+
+
+Packages are not just for functions and data-- you can create a package to deliver interactive tutorials to an audience, too!
+
+If you want to share your [learnr tutorial](https://rstudio.github.io/learnr/) with a large group of users (e.g. large courses where many students will likely run the same tutorials simultaneously), putting your tutorial inside of a package may actually be one the *best* ways to get your content to your audience. Why? Because after the package has been installed and loaded, users can run your tutorial *locally*-- which means that *you* don't have to worry about potential bandwidth limits and issues that might otherwise arise when learnr tutorials are hosted on external servers.
+
+Putting your learnr tutorial in a package is also free-- and always will be, which means you can also sidestep costs associated with hosting services! Your package doesn't have to be on CRAN, and it doesn't have to contain functions or data (but it can if you want!).
+
+Never made a package before? No sweat! In this post, we'll guide you through the process.
+
+
+
+## tl;dr
+
+Here's how you put an interactive learnr tutorial in a package:
+
+1. In a GitHub-connected RStudio project, run `usethis::create_package(getwd())` to make a basic package.
+
+1. Run `usethis::use_tutorial("<name-of-learnr-file>", "<Title You'd Like the User to See>")` to create the learnr tutorial. Customize with your content. Repeat for additional tutorials.
+
+1. Go to *Build* > *Install and Restart*.
+
+1. Optionally, run `usethis::use_package("gradethis")` to add packages dependencies for exercise-checking.
+
+1. Optionally, edit the `DESCRIPTION` file.
+
+1. Run `usethis::use_readme_rmd()` to add a README. Knit when you're done.
+
+1. Push to GitHub and have folks install with `remotes::install_github("<your-repo>/<package-name>")`
+
+1. Repeat steps 2, 3, and 7 anytime you update your tutorial package.  
+
+
+```{r, out.width = "100%", echo = FALSE}
+knitr::include_graphics("pkg-illo.jpg")
+```
+
+## How to make a package for a learnr tutorial
+
+
+We will walk through the steps for creating the bare-minimum package that you can use to house and deliver your interactive learnr tutorials to your users.
+
+Our final goal is to create a package that is "downloadable" by users, which means this package needs a GitHub repository so that it can be easily shared. If students wanted to, they’d be able to find the source code and solutions in your GitHub repository (or, with some spelunking, in the folder that houses all their R packages locally). But this would take some extra effort on the part of students.
```

Comment:
Should we say that it "needs" GitHub? It is the most straightforward way but maybe softening the language to say something along the lines of "there are other ways too, but this is what we'll walk you through" is better? (I'm not willing to die on this hill, could easily be convinced to keep it simple)

Reply:
I like the idea of including this side note. What's the general term that I could use here: "...needs GitHub (or some other _source code repository?_) so that it can be easily shared"

Reply:
How about...

> "Our final goal is to create a package that is "downloadable" by users. The most straightforward way to do this is to host the package in a GitHub repository, however, it is also possible to achieve the same outcome using other hosting services like GitLab or by distributing the package source code via other means."